### ISSUE TITLE
No N:N association is created

### ISSUE THREAD

**Issue by Rasmussor** [demianrasko/Dynamics-365-Workflow-Tools#39]
I have created two custom entities; event and event participant to test if the workflow is useful to associate an event and a contact directly (N:N). I have created two workflows:
1. 1) That triggers on the event participant creation and copies the URL of the event and insert the value into a single text line field with URL format on a associated contact, 2) triggers the second worklow.
2. Runs on the contact with the custom command made by you. It uses the Event URL from the field on the contact.

![image](https://user-images.githubusercontent.com/32491875/34106305-feff4ab0-e3f7-11e7-88ef-69db1e540cea.png)

Everything seems to be running fine and both workflows are completed. However, no action seems to happen then the custom command of associating the N:N is called:

![image](https://user-images.githubusercontent.com/32491875/34106417-6625726e-e3f8-11e7-8d30-6803ecbba3e2.png)

The URL of the event in the field on the contact is correct and opens when clicked.

The test environment is a Dynamics 365 version 9 online trial. Workflow version: 1.0.37.0.

**Comment by Rasmussor** [comment]
Apologies for the Danish language in the SSs.

**Comment by Lyfey** [comment]
Hi Rasmussor,

I am also interested in this functionality.

For my clarity, can I confirm the following understanding:

Event is a custom entity. An event has 1:N relationship with Event Participant so an Event can have many Event Participants and an Event Participant has one Event. Event Participant also has a N:1 relationship with Contact, so an Event Participant has one Contact and a Contact has many Event Participants.

Event <1----N> Event Participant <N----1> Contact

This implies we can link Contact to Event in a N:N relationship, which is what you are trying to do.

The Relationship Name and the Relationship Entity Name are both new_event_contact (which would be the CRM default name [best practice would be to create a new Publisher with a prefix other than new by the way :) ]). You are creating a Dynamic URL to Event on Contact of type URL and called EventURL.

You have a workflow (called "Associate Contact to Event" [translated]) as a Child Process that is called by the first workflow and executes against Contact with the settings above.

Please let me know if I have misunderstood. The documentation at
[https://github.com/demianrasko/Dynamics-365-Workflow-Tools/blob/master/docs/Associate%20Entity.md](https://github.com/demianrasko/Dynamics-365-Workflow-Tools/blob/master/docs/Associate%20Entity.md) is a little confusing.

Lyfe

**Comment by Rasmussor** [comment]
Hi Demian,

That is all correct.

Mvh.
Rasmus Sørensen

Den 19. dec. 2017 kl. 06.01 skrev Lyfey <notifications@github.com<mailto:notifications@github.com>>:


Hi Rasmussor,

I am also interested in this functionality.

For my clarity, can I confirm the following understanding:

Event is a custom entity. An event has 1:N relationship with Event Participant so an Event can have many Event Participants and an Event Participant has one Event. Event Participant also has a N:1 relationship with Contact, so an Event Participant has one Contact and a Contact has many Event Participants.

Event <1----N> Event Participant Contact

This implies we can link Contact to Event in a N:N relationship, which is what you are trying to do.

The Relationship Name and the Relationship Entity Name are both new_event_contact (which would be the CRM default name [best practice would be to create a new Publisher with a prefix other than new by the way :) ]). You are creating a Dynamic URL to Event on Contact of type URL and called EventURL.

You have a workflow (called "Associate Contact to Event" [translated]) as a Child Process that is called by the first workflow and executes against Contact with the settings above.

Please let me know if I have misunderstood. The documentation at
https://github.com/demianrasko/Dynamics-365-Workflow-Tools/blob/master/docs/Associate%20Entity.md is a little confusing.

Lyfe

**Comment by demianrasko** [comment]
Hi Lyfe & Rasmus, 
If you have a custom entity called "Event Participant", then you don't need to use the Associate action, you can use the standard Create feature from Workflows. If this is the case, you created a custom entity called "Event Participant" with two Lookups fields (relationships) to Contact and to Event. In this case you create two N-1 relationshipts (and no one N-N of the standard relationships)
The associate action works only for standard N-N relationships, where you don't create any intersection entity, you just create the N-N relationship between the entities.

Please, tell me if this helps you. If you need more help, please detail me with screenshots the customizations and workflows, or send me the solution to test it with your customizations.

Thanks!

**Comment by Lyfey** [comment]
Hi Demian and merry christmas! Thanks very much for your help and time over the holidays!

In Rasmus case above, you are correct in understanding that he has created an intermediate entity between Contact and Event. However, he wishes to _also_ have an N:N relationship between Contact and Event _as well_. My understanding is that he is creating a workflow to create the N:N relationship on create of the intermediate 'event participant' entity. Therefore he does want to create an N:N, but we are both confused by the use of the Associate action. Could you please describe the correct way to configure the action?

Could you please describe the use of:
- Relationship Name
- Relationship Entity Name
- Record URL

Lets assume the workflow is running against 'Event Participant A' and we wish to create an N:N relationship between 'Contact A' and Event record 'Event A'. We discover 'Contact A' and 'Event A' as they both are related to 'Event Participant A' as N:1 records (meaning Event Participant has 1 Contact and 1 Event).

Many thanks!

**Comment by demianrasko** [comment]
Hi, Let me try to explain the 2 scenarios.
1) if you created a new custom entity called "Event Participant" with two Lookups to Contact And Event, you don't need to create the Associate action, just use the standard Create action.
2) if you want to use the N-N relationships between Contact and Event entities, firs you need to create that relationship in the entity here (you can create it from Contact or from Event entities):
![createn-n](https://user-images.githubusercontent.com/18124945/34336575-96848246-e958-11e7-93b3-2f085e4747c4.png)
Once you have created the N-N relationship, you will have these values to be mapped as explained in the documentation:
![fieldsmapping](https://user-images.githubusercontent.com/18124945/34336607-ddf3cf6a-e958-11e7-845e-3fa5f26fead0.png)

I think maybe you want to create the N-N relationship, but you didn't created the relationship before.

Hope this helps.

**Comment by Rasmussor** [comment]
Hi Demian and Lyfey,

I have already created the N:N between contacts and events. 

The reason why I would like a N:N relationship is for the end users to easily access the contacts associated with an event from the event form. Hereby contacts can be added to marketing lists and emails can be send out to them. 

The reason I also use an intermediary entity like 'event participant' is for participants to be able to sign up from a website creating 'event participants' (which in fact would just be webforms), and contacts if they do not already exist in the system. These contacts should be added to the event as described previously.

However, the action does not execute of creating the N:N relationship. Please let me know, if you want any further specifications. 

Thanks.

**Comment by Lyfey** [comment]
Hi Rasmussor,

I've recreated your scenario on a test box and have it working. I completed it using two workflows, but in a different way.

Lets assume the workflow is running against 'Event Participant A' and we wish to create an N:N relationship between 'Contact A' and Event record 'Event A'. We discover 'Contact A' and 'Event A' as they both are related to 'Event Participant A' as N:1 records (meaning Event Participant has 1 Contact and 1 Event).

On Contact, I created a new N:1 relationship (many Contacts may have 1 Event) between Contact and Event called "Event to be added". On the first workflow, I associated Contact A to Event A in this field. This field is not displayed on the form.

I have a second workflow which triggers onChange of "Event to be added". If this field contains data, it calls the Associate Entity custom assembly. Relationship Name and Relationship Entity Name are new_contact_event. Record URL is "Event to be added"/Record URL(Dynamic). It then clears the "Event to be added" field for cleanliness.

I think the part that is missing from Demians documentation is that the assembly gets its reference to the first entity from the context of the entity the workflow is executing against.

In short:
"Primary" N:N entity must be the entity the workflow calling the action.
Relationship Name and Relationship Entity Name : This is the schema name of the N:N relation, as per Demians documentation
Record URL: is the URL to the record to be associated with, in the format
https://[server]/main.aspx?etc=[type]&id=[guid]&histKey=33902444&newWindow=true&pagetype=entityrecord
The easiest way to generate that is via Record URL (Dynamic) as Demian suggests, and that usually needs the record associated already, hence the "Event to be added" field.

Hope this helps!

**Comment by demianrasko** [comment]
Hi Lyfey, thats correct, probbably this is the point. Thanks for explaining so good.
@Rasmussor : please share me the solution with the entities, and I will try to response you with the solution with the process included working.

Thanks!

**Comment by Rasmussor** [comment]
Hi Lyfey and Demian

Thanks for your responses and help. I really appreciate it! 

I tried to create Lyfey's suggestions for a solution in a different environment, but still v. 9. Everything seems to work, but the execution of the plugin / custom demand. So the same result as before unfortunately. 

I have attached the solution to this comment. It is in English. Please feel free to review. 

[Eventmanagement_1_0_0_1.zip](https://github.com/demianrasko/Dynamics-365-Workflow-Tools/files/1635128/Eventmanagement_1_0_0_1.zip)

**Comment by demianrasko** [comment]
Hi, I tried to import in a clead environment, but there are some missing dependencies...
Look at the error:
![image](https://user-images.githubusercontent.com/18124945/35483144-f2313606-043e-11e8-8a11-e6677dbf0c0b.png)

**Comment by easybeat** [comment]
@Rasmussor 
I'm facing some problems with the settings too. Could you please explain in my case what I have to put in which setting fields?

My scenario:
I have an Entity "Unsubscribe" on which I want to run the workflow. "Unsubscribe" has an N:N relation to "Subscription List". Please fill for me:

Relationsship Name:
Relationsship Entity Name:
Record URL:

Thank you very much for your help!
Kind regards
Beat
